packages/web5/lib/src/pexv2/pd.dart

```diff
@@ -0,0 +1,274 @@
+      InputDescriptor(
+        id: json['id'],
+        name: json['name'],
+        purpose: json['purpose'],
+        constraints: Constraints.fromJson(json['constraints']),
+      );
+
+  Map<String, dynamic> toJson() => {
+        'id': id,
+        'name': name,
+        'purpose': purpose,
+        'constraints': constraints.toJson(),
+      };
+
+  String _generateRandomToken() {
+    final rand = Random.secure();
+    final bytes = Uint8List(16);
+    for (int i = 0; i < 16; i++) {
+      bytes[i] = rand.nextInt(256);
+    }
+    return hex.encode(bytes);
+  }
+
+  List<String> selectCredentials(List<String> vcJWTs) {
+    final List<String> answer = [];
+    final List<_TokenizedField> tokenizedField = [];
+    final schemaMap = {
+      '\$schema': 'http://json-schema.org/draft-07/schema#',
+      'type': 'object',
+      'properties': {},
+      'required': [],
+    };
+
+    // Populate JSON schema and generate tokens for each field
+    for (var field in constraints.fields ?? []) {
+      final token = _generateRandomToken();
+      tokenizedField
+          .add(_TokenizedField(token: token, paths: field.path ?? []));
+
+      final properties = schemaMap['properties'] as Map<String, dynamic>;
+
+      if (field.filter != null) {
+        properties[token] = field.filter.toJson();
+      } else {
+        final anyType = {
+          'type': ['string', 'number', 'boolean', 'object', 'array'],
+        };
+        properties[token] = anyType;
+      }
+      final required = schemaMap['required'] as List<dynamic>;
+      required.add(token);
+    }
+    final jsonSchema = JsonSchema.create(schemaMap);
+
+    // Tokenize each vcJwt and validate it against the JSON schema
+    for (var vcJWT in vcJWTs) {
+      final decoded = json.decode(vcJWT);
+
+      final selectionCandidate = <String, dynamic>{};
+
+      for (var tokenPath in tokenizedField) {
+        for (var path in tokenPath.paths) {
+          final value = JsonPath(path)
+              .read(decoded)
+              .firstOrNull; // Custom function needed to handle JSON paths.
+          if (value != null) {
+            selectionCandidate[tokenPath.token] = value;
+            break;
+          }
+        }
+      }
+
+      final validationResult = jsonSchema.validate(selectionCandidate);
+      if (validationResult.isValid) {
+        answer.add(vcJWT);
+      }
+    }
+
+    return answer;
+  }
+}
+
+/// Constraints contains the requirements for a given Input Descriptor.
+class Constraints {
+  List<Field>? fields;
+
+  Constraints({this.fields});
+
+  factory Constraints.fromJson(Map<String, dynamic> json) => Constraints(
+        fields: json['fields'] == null
+            ? null
+            : List<Field>.from(json['fields'].map((x) => Field.fromJson(x))),
+      );
+
+  Map<String, dynamic> toJson() => {
+        'fields': fields == null
+            ? null
+            : List<dynamic>.from(fields!.map((x) => x.toJson())),
+      };
```

Comment:
indentation is interesting here

Reply:
agreed it's a lil weird. this is the indentation that `dart fix` seems to want though